File: src/main/java/io/vproxy/vfx/entity/input/InputData.java
package io.vproxy.vfx.entity.input;

import javafx.scene.input.MouseButton;
import vjson.JSON;
import vjson.deserializer.rule.BoolRule;
import vjson.deserializer.rule.ObjectRule;
import vjson.deserializer.rule.Rule;
import vjson.deserializer.rule.StringRule;
import vjson.util.ObjectBuilder;

import java.util.Set;

public class InputData {
    public boolean ctrl;
    public boolean alt;
    public boolean shift;
    public Key key;

    public static final Rule<InputData> rule = new ObjectRule<>(InputData::new)
        .put("ctrl", (o, it) -> o.ctrl = it, BoolRule.get())
        .put("alt", (o, it) -> o.alt = it, BoolRule.get())
        .put("shift", (o, it) -> o.shift = it, BoolRule.get())
        .put("key", (o, it) -> o.key = new Key(it), StringRule.get());

    public InputData() {
    }

    public InputData(InputData data) {
        this.ctrl = data.ctrl;
        this.alt = data.alt;
        this.shift = data.shift;
        this.key = data.key;
    }

    public InputData(boolean ctrl, boolean alt, boolean shift, Key key) {
        this.ctrl = ctrl;
        this.alt = alt;
        this.shift = shift;
        this.key = key;
    }

    public InputData(Key key) {
        this(false, false, false, key);
    }

    public boolean matches(Set<KeyCode> keys, Set<MouseButton> buttons, KeyCode currentKey, MouseButton currentMouse) {
        if (ctrl) {
            if (!keys.contains(KeyCode.CONTROL)) return false;
        }
        if (alt) {
            if (!keys.contains(KeyCode.ALT)) return false;
        }
        if (shift) {
            if (!keys.contains(KeyCode.SHIFT)) return false;
        }
        if (key.key != null) return key.key == currentKey;
        if (key.button != null) return key.button == currentMouse;
        if (currentKey == KeyCode.CONTROL) return ctrl;
        if (currentKey == KeyCode.ALT) return alt;
        if (currentKey == KeyCode.SHIFT) return shift;
        return false;
    }

    public JSON.Object toJson() {
        var ob = new ObjectBuilder()
            .put("ctrl", ctrl)
            .put("alt", alt)
            .put("shift", shift);
        if (key != null) {
            ob.put("key", key.toString());
        }
        return ob.build();
    }
}

File: src-ts/src/entity/input/InputData.ts
import { Key, MouseButton } from './Key.js';
import { KeyCode, keyCodeByUeText } from './KeyCode.js';

export class InputData {
  ctrl: boolean;
  alt: boolean;
  shift: boolean;
  key: Key | null;

  constructor();
  constructor(data: InputData);
  constructor(key: Key);
  constructor(ctrl: boolean, alt: boolean, shift: boolean, key: Key);
  constructor(...args: unknown[]) {
    if (args.length === 0) {
      this.ctrl = false;
      this.alt = false;
      this.shift = false;
      this.key = null;
    } else if (args.length === 1 && args[0] instanceof InputData) {
      const d = args[0] as InputData;
      this.ctrl = d.ctrl;
      this.alt = d.alt;
      this.shift = d.shift;
      this.key = d.key;
    } else if (args.length === 1 && args[0] instanceof Key) {
      this.ctrl = false;
      this.alt = false;
      this.shift = false;
      this.key = args[0];
    } else {
      this.ctrl = args[0] as boolean;
      this.alt = args[1] as boolean;
      this.shift = args[2] as boolean;
      this.key = args[3] as Key;
    }
  }

  matches(
    keys: Set<KeyCode>,
    buttons: Set<MouseButton>,
    currentKey: KeyCode | null,
    currentMouse: MouseButton | null,
  ): boolean {
    const ctrlCode = keyCodeByUeText('control');
    const altCode = keyCodeByUeText('alt');
    const shiftCode = keyCodeByUeText('shift');
    if (this.ctrl && !(ctrlCode && keys.has(ctrlCode))) return false;
    if (this.alt && !(altCode && keys.has(altCode))) return false;
    if (this.shift && !(shiftCode && keys.has(shiftCode))) return false;
    // Java's matches() assumes `key` is non-null when invoked; we keep the
    // same assumption but guard for the null no-arg-ctor case.
    if (this.key !== null && this.key.key !== null) return this.key.key === currentKey;
    if (this.key !== null && this.key.button !== null) return this.key.button === currentMouse;
    if (currentKey && ctrlCode && currentKey === ctrlCode) return this.ctrl;
    if (currentKey && altCode && currentKey === altCode) return this.alt;
    if (currentKey && shiftCode && currentKey === shiftCode) return this.shift;
    return false;
  }

  toJson(): Record<string, unknown> {
    const obj: Record<string, unknown> = {
      ctrl: this.ctrl,
      alt: this.alt,
      shift: this.shift,
    };
    if (this.key !== null) obj.key = this.key.toString();
    return obj;
  }
}
